port/src/devicetree.rs

```diff
@@ -0,0 +1,738 @@
+use core::ffi::CStr;
+
+#[derive(Debug)]
+pub enum ParseError {
+    InvalidHeader,
+    InvalidMagic,
+    BufferTooSmall,
+    InvalidToken,
+}
+
+type Result<T> = core::result::Result<T, ParseError>;
+
+fn bytes_to_u32(bytes: &[u8]) -> Option<u32> {
+    Some(u32::from_be_bytes(bytes.get(..4)?.try_into().unwrap()))
+}
+
+fn bytes_to_u64(bytes: &[u8]) -> Option<u64> {
+    Some(u64::from_be_bytes(bytes.get(..8)?.try_into().unwrap()))
+}
+
+/// Extract u32 from bytes, but cast as u64
+fn bytes_to_u32_as_u64(bytes: &[u8]) -> Option<u64> {
+    Some(u32::from_be_bytes(bytes.get(..4)?.try_into().unwrap()) as u64)
+}
+
+fn align4(n: usize) -> usize {
+    let mut aligned_n = n;
+    let rem = n % 4;
+    if rem > 0 {
+        aligned_n += 4 - rem;
+    }
+    aligned_n
+}
+
+/// DeviceTree is the class entrypoint to the Devicetree operations.
+/// This code focuses only on parsing a Flattened Devicetree without using the heap.
+/// The Devicetree specification can be found here:
+/// https://www.devicetree.org/specifications/
+#[derive(Debug)]
+pub struct DeviceTree<'a> {
+    data: &'a [u8],    // Reference to the underlying data in memory
+    header: FdtHeader, // Parsed structure of the header
+}
+
+impl<'a> DeviceTree<'a> {
+    /// Create new DeviceTree based on memory pointed to by data.
+    /// Result is error if the header can't be parsed correctly.
+    pub fn new(data: &'a [u8]) -> Result<Self> {
+        FdtHeader::new(data, false).map(|header| Self { data, header })
+    }
+
+    /// Given a pointer to the dtb as a u64, return a DeviceTree struct.
+    pub fn from_u64(ptr: u64) -> Result<Self> {
+        let u8ptr = ptr as *const u8;
+
+        // Extract the real length from the header
+        let dtb_buf_for_header: &[u8] = unsafe { core::slice::from_raw_parts(u8ptr, 100) };
```

Comment:
Why 100? What if that goes beyond the end of some mapped region?

Reply:
There are also sound memory soundness issues here; the first is there's no guarantee that there is valid data here at all; the compiler might poison this: I think you may want to do a, `MaybeUninit` dance here. The second is that you've got no guarantee that another Rust object won't alias this or part of it mutably.

Reply:
So presumably that would mean treating the &[u8] as &[MaybeUninit<u8>] throughout the fdt code?

Reply:
Having looked a bit more, are you suggesting we treat all &[u8] in this code as &[MaybeUninit<u8>] until as late as possible?

Reply:
Ok @dancrossnyc - not sure if this is what you had in mind, but check out the latest commit.  If there's a particular problem with how I did it (I've never used MaybeUninit before), let me know, or even better, provide a fix - otherwise we might go round in circles :)  Thanks.

Reply:
Any feedback here @dancrossnyc ?  Would be good to close this.